### DISCUSSION TITLE
Instrumenting binaries issue

### DISCUSSION THREAD

**Discussion by Piotr-Miller** [SteveGilham/altcover#155] [Q&A] [answered]
Hi Guys,

Got issue with instrumenting binaries on build server (Win Server 2019 Standard) - couple days ago altcover stopped seeing all of my dlls and see only third party ones: CefSharp.Core.dll, Ben.Demystifier.dll, etc, which I need filter out anyway.

After running command:
     **altcover --inplace -i "Path to release folder" -s "xxxx" --single**
coverage.xml file looks like below:

```
<?xml version="1.0" encoding="utf-8" standalone="yes"?>
<CoverageSession xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
  <Summary numSequencePoints="0" visitedSequencePoints="0" numBranchPoints="0" visitedBranchPoints="0" sequenceCoverage="0" branchCoverage="0" maxCyclomaticComplexity="0" minCyclomaticComplexity="1" visitedClasses="0" numClasses="0" visitedMethods="0" numMethods="0" minCrapScore="0" maxCrapScore="0" />
  <Modules />
</CoverageSession>
```

First of my thoughts was something wrong happend with binaries, but on my local machine (Windows 10 Enterprise) all is working flawlessly:

```
<?xml version="1.0" encoding="utf-8" standalone="yes"?>
<CoverageSession xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
  <Summary numSequencePoints="50858" visitedSequencePoints="0" numBranchPoints="32610" visitedBranchPoints="0" sequenceCoverage="0" branchCoverage="0" maxCyclomaticComplexity="358" minCyclomaticComplexity="1" visitedClasses="0" numClasses="1609" visitedMethods="0" numMethods="10730" minCrapScore="0" maxCrapScore="0" />
  <Modules>
    <Module hash="63-AF-83-07-9C-67-6A-43-0A-7F-E0-A3-E3-9C-CB-4D-1B-B8-A5-02">
      <Summary numSequencePoints="0" visitedSequencePoints="0" numBranchPoints="0" visitedBranchPoints="0" sequenceCoverage="0" branchCoverage="0" maxCyclomaticComplexity="0" minCyclomaticComplexity="1" visitedClasses="0" numClasses="0" visitedMethods="0" numMethods="0" minCrapScore="0" maxCrapScore="0" />
      <ModulePath>C:\xxx\__Saved\xxx.dll</ModulePath>
      <ModuleTime>2022-06-02T11:01:34.1420363Z</ModuleTime>
      <ModuleName>xxx</ModuleName>
      <Files />
      <Classes>
        <Class>
          <Summary numSequencePoints="0" visitedSequencePoints="0" numBranchPoints="0" visitedBranchPoints="0" sequenceCoverage="0" branchCoverage="0" maxCyclomaticComplexity="0" minCyclomaticComplexity="0" visitedClasses="0" numClasses="0" visitedMethods="0" numMethods="0" minCrapScore="0" maxCrapScore="0" />
          <FullName>yyyyy</FullName>
          ...
```

I'm using on both locations (my dev machine and build server) the same set of dlls, the same version of altcover.

I'm realising that most probable source of this issue is related with build server configuration, not altcover itself,
but decided to ask you guys anyway.

Am I missing something? 

Thanks in advance for your support.

BR,
Piotr Miller

**Comment by SteveGilham** [comment, marked as answer]
The most obvious cause for an assembly to be skipped is that matching debug symbols  -- the link between the IL in the assembly and the source code being covered -- are not being found.  It is even possible that they are not being generated at all on the build server, as they are switched off by default for the Release build configuration.

The other assemblies you mention do come packaged with embedded debug information -- like this
![Screenshot 2022-06-08 092750](https://user-images.githubusercontent.com/7460854/172569823-a4d2e789-1483-4715-b064-af1360e12dee.png)
which is the first place the instrumentation process looks.  Next is the symbol file location from the assembly debug header (typically, the intermediate build folder), next is in the same directory as the assembly, next are any folders listed as `--symbolDirectory` parameters, stopping at the first suitably named file found.

**Reply by Piotr-Miller** [reply]
Hi Steve,

All necessary pdb files are generated and sit in the same folder as assemblies.
When I completely remove all filtering the only instrumented files are third party ones, so filtering is not the case (moreover I've added --symbolDirectory parameter as well): 

altcover --inplace -i E:\xxx -y E:\xxx --single

Creating folder E:\xxx\__Saved
Saving files to E:\xxx\__Saved
Instrumenting files in E:\xxx
   => E:\xxx\__Saved\Ben.Demystifier.dll
   => E:\xxx\__Saved\CefSharp.BrowserSubprocess.dll
   => E:\xxx\__Saved\CefSharp.Core.dll
   => E:\xxx\__Saved\CefSharp.dll
   => E:\xxx\__Saved\CefSharp.Wpf.dll
   => E:\xxx\__Saved\System.Reactive.dll

Coverage Report: C:\yyyy\coverage.xml

    E:\xxx\System.Reactive.dll
                <=  System.Reactive, Version=5.0.0.0, Culture=neutral, PublicKeyToken=null
    E:\xxx\CefSharp.dll
                <=  CefSharp, Version=96.0.180.0, Culture=neutral, PublicKeyToken=null
    E:\xxx\Ben.Demystifier.dll
                <=  Ben.Demystifier, Version=0.4.0.0, Culture=neutral, PublicKeyToken=null
    E:\xxx\CefSharp.Core.dll
                <=  CefSharp.Core, Version=96.0.180.0, Culture=neutral, PublicKeyToken=null
    E:\xxx\CefSharp.BrowserSubprocess.dll
                <=  CefSharp.BrowserSubprocess, Version=96.0.180.0, Culture=neutral, PublicKeyToken=null
    E:\xxx\CefSharp.Wpf.dll
                <=  CefSharp.Wpf, Version=96.0.180.0, Culture=neutral, PublicKeyToken=null

And as I said locally with the same altcover version and the same package of assemblies I have no issue, pdb files are found and all necessary binaries are instrumented. This is weird.
I thought that someone could faced  similar behaviour.

**Reply by SteveGilham** [reply]
I'm not sure how many users of altcover scan these discussions, so it may only be me here to answer.

Another class of assemblies that get skipped over are ones that have already been instrumented, but it would take a highly unlikely set of circumstances to leave those lying around.  It would explain a "the coverage suddenly stopped working" scenario, but amongst other things, would require new builds to fail to clean or overwrite the output directory following a mid-test crash during an earlier run.  Simpler circumstances like multiple concurrent coverage tests being run for a shared build output directory would let at least one test per build instrument the files.

Unfortunately, this is a part of the process where there is currently no logging set up to say why an assembly hasn't been processed.  It looks like there's an implicit enhancement request here.

**Reply by Piotr-Miller** [reply]
Yep, for sure it would help investigate some issues. ;)

**Reply by SteveGilham** [reply]
OK, here's a prototype, fresh out of the GitHub CI build.  It adds a `--verbose` command line argument that outputs extra details about what's going on.  So far, the added output traces files going into the process of being instrumented; this is a proof of concept, and the tracing can be extended further in due course.

[altcover.8.3.653-github-pre.nupkg.zip](https://github.com/SteveGilham/altcover/files/8873303/altcover.8.3.653-github-pre.nupkg.zip)
.
Basic package in a .zip archive to placate GitHub's limits on embeds.

**Reply by Piotr-Miller** [reply]
I did tests on build server and my local machine with parameter --verbose.

Build server:

      E:\xxx.Build189\__Saved\zzz.Connectivity.dll : beginning process
      E:\xxx.Build189\__Saved\zzz.Connectivity.dll : ** not a valid assembly **
      E:\xxx.Build189\__Saved\zzz.Connectivity.pdb : beginning process
      E:\xxx.Build189\__Saved\zzz.Connectivity.pdb : ** not a valid assembly **
      E:\xxx.Build189\__Saved\zzz.Fonts.dll : beginning process
      E:\xxx.Build189\__Saved\zzz.Fonts.dll : ** not a valid assembly **
      E:\xxx.Build189\__Saved\zzz.Fonts.pdb : beginning process
      E:\xxx.Build189\__Saved\zzz.Fonts.pdb : ** not a valid assembly **

      Coverage Report: C:\Users\pmiller\coverage.xml

Local machine:

      E:\xxx.Build189\__Saved\zzz.Connectivity.dll : beginning process
         => E:\xxx.Build189\__Saved\zzz.Connectivity.dll
      E:\xxx.Build189\__Saved\zzz.Connectivity.pdb : beginning process
      E:\xxx.Build189\__Saved\zzz.Connectivity.pdb : ** not a valid assembly **
      E:\xxx.Build189\__Saved\zzz.Fonts.dll : beginning process
         => E:\xxx.Build189\__Saved\zzz.Fonts.dll
      E:\xxx.Build189\__Saved\zzz.Fonts.pdb : beginning process
      E:\xxx.Build189\__Saved\zzz.Fonts.pdb : ** not a valid assembly **
      
      Coverage Report: C:\Users\pmiller\coverage.xml
      
          E:\xxx.Build189\zzz.Fonts.dll
                      <=  zzz.Fonts, Version=1.0.8193.24506, Culture=neutral, PublicKeyToken=null
          E:\xxx.Build189\zzz.Connectivity.dll
                      <=  zzz.Connectivity, Version=1.0.8193.24506, Culture=neutral, PublicKeyToken=null

Processing the same assemblies on build server resulting message ** not a valid assembly **.
Weird.

**Reply by SteveGilham** [reply]
That comes from the point where files are being opened for read, and inspected for symbols, and conflates a number of exception cases that would fail a file, but not the whole process.  While it can mean "this is not an assembly", as can be seen from processing the `.pdb` files, the most likely cause would be `Mono.Cecil.Cil.SymbolsNotMatchingException`, with the process picking up symbols from a file of the expected name, but from a previous or subsequent compilation.

Tracing symbol resolution would be a useful thing to add to the verbose mode, along with disambiguating the exceptions at this point.

**Reply by SteveGilham** [reply]
Here's a build which should record separately the reasons why a file is not valid, and some further details of the instrumentation process, so may shed some light on the issue 

[altcover.8.3.660-github-pre.nupkg.zip](https://github.com/SteveGilham/altcover/files/8930842/altcover.8.3.660-github-pre.nupkg.zip)

Working through this part of the code I have spotted one part where ambiguities -- two or more possible symbol files found by name -- are resolved arbitrarily, rather than by inspecting in detail whether they match the assembly in question, so I shall be addressing that as well.

**Reply by Piotr-Miller** [reply]
Hi Steve,

You were right, process on build server is trying to pick up symbol files from subsequent compilation:
	**Assembly E:\xxx.Build189\__Saved\zzz.Connectivity.dll symbols from image 'E:\GitLab_builds_folder\net6.0-windows10.0.17763.0\win-x64\zzz.Connectivity.pdb'**
On my dev machine process is using __Saved folder, and both binary file and symbol are matched:
	**Assembly E:\xxx.Build189\__Saved\zzz.Connectivity.dll symbols from folder '"E:\xxx.Build189\__Saved\zzzz.Connectivity.pdb"'**

Build server:

E:\xxx.Build189\__Saved\zzz.Connectivity.dll : beginning process
Assembly E:\xxx.Build189\__Saved\zzz.Connectivity.dll symbols from image 'E:\GitLab_builds_folder\net6.0-windows10.0.17763.0\win-x64\zzz.Connectivity.pdb'
E:\xxx.Build189\__Saved\zzz.Connectivity.dll : ** not a valid assembly because Mono.Cecil.Cil.SymbolsNotMatchingException: Symbols were found but are not matching the assembly
   at Mono.Cecil.ModuleDefinition.ReadSymbols(ISymbolReader reader, Boolean throwIfSymbolsAreNotMaching)
   at AltCover.ProgramDatabase.readSymbols@106.Invoke(String pdbpath) in /_//AltCover.Engine/ProgramDatabase.fs:line 116
   at AltCover.Main.I.assemblies@543-4.Invoke(Unit unitVar0) in /_//AltCover.Engine/Main.fs:line 551
   at AltCover.Main.I.imageLoadResilient[a](FSharpFunc`2 f, FSharpFunc`2 tidy) in /_//AltCover.Engine/Main.fs:line 404 **
E:\xxx.Build189\__Saved\zzz.Connectivity.pdb : beginning process
E:\xxx.Build189\__Saved\zzz.Connectivity.pdb : ** not a valid assembly because System.BadImageFormatException: Format of the executable (.exe) or library (.dll) is invalid. **

Coverage Report:  C:\Users\pmiller\coverage.xml

Local machine:

E:\xxx.Build189\__Saved\zzz.Connectivity.dll : beginning process
Assembly E:\xxx.Build189\__Saved\zzz.Connectivity.dll symbols from folder '"E:\xxx.Build189\__Saved\zzzz.Connectivity.pdb"'
   => E:\xxx.Build189\__Saved\zzz.Connectivity.dll
E:\xxx.Build189\__Saved\zzz.Connectivity.pdb : beginning process
E:\xxx.Build189\__Saved\zzz.Connectivity.pdb : ** not a valid assembly because System.BadImageFormatException: Format of the executable (.exe) or library (.dll) is invalid. **

Coverage Report: C:\coverage.xml

Assembly  symbols from image 'AltCover.Recorder.pdb'
Assembly E:\xxx.Build189\__Saved\zzz.Connectivity.dll symbols from folder '"E:\xxx.Build189\__Saved\zzz.Connectivity.pdb"'
    E:\xxx.Build189\zzz.Connectivity.dll
                <=  zzz.Connectivity, Version=1.0.8193.24506, Culture=neutral, PublicKeyToken=null

For build server passing explicitly --symbolDirectory parameter as E:\xxx.Build189 did not do the job, unfortunately.

BTW, huge thanks for your support.

**Reply by SteveGilham** [reply]
No problem.

The code as it stands just works through a list of locations - the one specified in the debug header, co-located with the assembly, then each of the `--symbolDirectory` values, stopping as soon as it finds one with the expected name, on the -- now invalidated -- assumption that there would not be cross-contamination from other builds which might be found first.  As Mono.Cecil doesn't give a convenient API for doing this, I'm in the process of writing code to minimally interpret a `.pdb` file to check whether it has the same GUID as in the assembly debug header, as well as simply having the expected name.

For the moment, the work-round is to change the build, at least for the configuration under test, to generate embedded `.pdb` information
```xml
    <DebugType>embedded</DebugType>
```
which puts the debug information in the first place that the process looks, and avoids all the problems involved with separate files.

**Reply by Piotr-Miller** [reply]
Thanks Steve, now it's clear for me what is going on behind the scene.
It would be good to force altcover to search symbols in input directory firstly.

**Reply by SteveGilham** [reply]
Here's a build for you to evaluate where I've reworked the symbol file selection, and should resolve the original underlying issue

[altcover.8.3.677-github-pre.nupkg.zip](https://github.com/SteveGilham/altcover/files/8986445/altcover.8.3.677-github-pre.nupkg.zip)

**Reply by SteveGilham** [reply]
Now in release v8.3.838
